src/core/namespace.c

```diff
@@ -1851,54 +1883,93 @@ int setup_netns(const int netns_storage_socket[static 2]) {
+        if (r < 0)
+                return r;
 
         netns = receive_one_fd(netns_storage_socket[0], MSG_DONTWAIT);
         if (netns == -EAGAIN) {
                 /* Nothing stored yet, so let's create a new namespace. */
 
-                if (unshare(CLONE_NEWNET) < 0) {
-                        r = -errno;
-                        goto fail;
-                }
+                if (unshare(CLONE_NEWNET) < 0)
+                        return -errno;
 
                 (void) loopback_setup();
 
                 netns = open("/proc/self/ns/net", O_RDONLY|O_CLOEXEC|O_NOCTTY);
-                if (netns < 0) {
-                        r = -errno;
-                        goto fail;
-                }
+                if (netns < 0)
+                        return -errno;
 
                 r = 1;
 
-        } else if (netns < 0) {
-                r = netns;
-                goto fail;
+        } else if (netns < 0)
+                return netns;
 
-        } else {
+        else {
                 /* Yay, found something, so let's join the namespace */
-                if (setns(netns, CLONE_NEWNET) < 0) {
-                        r = -errno;
-                        goto fail;
-                }
+                if (setns(netns, CLONE_NEWNET) < 0)
+                        return -errno;
 
                 r = 0;
         }
 
         q = send_one_fd(netns_storage_socket[1], netns, MSG_DONTWAIT);
-        if (q < 0) {
-                r = q;
-                goto fail;
-        }
+        if (q < 0)
+                return q;
 
-fail:
-        (void) lockf(netns_storage_socket[0], F_ULOCK, 0);
         return r;
 }
 
+static int netns_create(const char *path, bool create) {
+        _cleanup_close_ int old_netns = -1, netns = -1;
+        int r;
+
+        if (create) {
+                r = mkdir_parents(path, 0755);
+                if (r < 0)
+                        return r;
+
+                r = touch(path);
+                if (r < 0)
+                        return r;
+        }
+
+        /* Save the current netns. */
+        old_netns = open("/proc/self/ns/net", O_RDONLY | O_CLOEXEC);
+        if (old_netns < 0)
+                return -errno;
+
+        /* Create a new netns */
+        if (unshare(CLONE_NEWNET) < 0)
+                return -errno;
+
+        (void) loopback_setup();
+
+        /* Mount the new netns onto the path. */
+        if (mount("/proc/self/ns/net", path, "none", MS_BIND, NULL) < 0) {
+                r = -errno;
+                (void) setns(old_netns, CLONE_NEWNET);
+                return r;
+        }
+
+        /* Open the new netns. */
+        netns = open("/proc/self/ns/net", O_RDONLY | O_CLOEXEC);
+        if (netns < 0) {
+                r = -errno;
+                (void) umount(path);
+                (void) setns(old_netns, CLONE_NEWNET);
+                return r;
+        }
+
+        /* Restore the original netns. */
+        if (setns(old_netns, CLONE_NEWNET) < 0)
```

Comment:
I know @poettering asked me to do namespace changes in a child process to make cleanup easier in an nspawn PR review. I'm not sure if this is a code path where we can afford to fork a child process but if we can, it might be easier to fork, unshare and mount instead of having logic everywhere to move back to the original network namespace.

Reply:
I will re-consider this point later. Thanks.

Reply:
yes, as mentioned elsewhere, let's do that in a child process. i.e. clone() with CLONE_NEWNET, and pass netns fd up to parent again